File: python/dungeon-game.py
# Using bottom up dp approach to calculate minimum cost path

class Solution:
    def calculateMinimumHP(self, grid):
        m = len(grid)
        n = len(grid[0])
        dp = [[0 for i in range(n)]for j in range(m) ]
        dp[-1][-1] = abs(grid[-1][-1])+1 if grid[-1][-1]<0 else 1

        # For last column
        for i in range(m-2,-1,-1):
            dp[i][-1] = max(dp[i+1][-1] - grid[i][n-1],1)

        # for last row
        for i in range(n-2,-1,-1):
            dp[-1][i] = max(dp[-1][i+1]-grid[-1][i],1)

        # for rest of the dp
        for i in range(m-2,-1,-1):
            for j in range(n-2,-1,-1):
                dp[i][j] = max(min(dp[i+1][j],dp[i][j+1])-grid[i][j],1)
        

        return dp[0][0]

File: java/dungeon-game.java
class Solution {
    public int calculateMinimumHP(int[][] dungeon) {
        int row = dungeon.length;
        int col = dungeon[0].length;
        int[][] dp = new int[row][col];
        
        
        //last element 
            dp[row-1][col-1] = dungeon[row-1][col-1]>0? 1:1-dungeon[row-1][col-1];

        // last row
            for(int i=col-2;i>=0;--i){
                dp[row-1][i] = Math.max(dp[row-1][i+1]-dungeon[row-1][i] ,1);

            }
        // last col
            for(int i = row-2;i>=0;--i){
                dp[i][col-1] = Math.max(dp[i+1][col-1]-dungeon[i][col-1],1);

            }

            //rest elements
            for(int i = row-2;i>=0;--i){
                for(int j = col-2;j>=0;--j){
                    dp[i][j] = Math.max(Math.min(dp[i+1][j],dp[i][j+1])-dungeon[i][j],1);

                }
            }
    return dp[0][0];
    }
}
